### PR TITLE
Add `with` statement feature

### PR DESCRIPTION
**Pull request by philipturner**
Allow ergonomic usage of Python context managers by mirroring `with` statements. This deviates from the current API because it adds an instance method to `PythonInterface`. Most built-in Python functions are accessible through syntax like `Python.open` and `Python.classmethod`, which uses `@dynamicMemberLookup` . The most reasonable way to incorporate `with` is by treating it as a built-in function, `Python.with`. An alternative is `PythonWith`, which does not look very pleasant.

Before this pull request, using context managers was very cumbersome. You had to do the following just to disable gradients in PyTorch:

```swift
// some code
do { // encapsulate the temporary variable's scope
  let no_grad = torch.no_grad()
  no_grad.__enter__()
  defer { no_grad.__exit__() }
  
  // some code
}
// some code
```

This is possible to mess up (by forgetting to exit the context manager) and obfuscates the code's semantic meaning. I introduced a much more elegant solution:

```swift
// some code
Python.with(torch.no_grad()) { _ in
  // some code
}
// some code
```

This also emulates the `with expression() as result:` syntax. You can use the expression's return value as the input to the closure.

```swift
Python.with(Python.open("temp", "w")) { opened_file in
    opened_file.write("Contents")
}
```

## Potential Problems

If Python already had a built-in function called `with`, which was called using `with()`, this would create a source break. [No such function exists](https://docs.python.org/3/library/functions.html) at this moment, so there is no API break.

### CODE REVIEW

**Comment by philipturner** [comment]
I could not get the CI to work on Windows. It's using a very old Swift toolchain (5.4.2), so perhaps it could be updated to 5.6.2. Please squash my commits before merging.
